File: release/src/router/samba-3.5.8/source3/lib/netapi/samr.c
/*
 *  Unix SMB/CIFS implementation.
 *  NetApi Samr Support
 *  Copyright (C) Guenther Deschner 2008
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, see <http://www.gnu.org/licenses/>.
 */

#include "includes.h"
#include "lib/netapi/netapi.h"
#include "lib/netapi/netapi_private.h"
#include "../librpc/gen_ndr/cli_samr.h"

/****************************************************************
****************************************************************/

WERROR libnetapi_samr_open_domain(struct libnetapi_ctx *mem_ctx,
				  struct rpc_pipe_client *pipe_cli,
				  uint32_t connect_mask,
				  uint32_t domain_mask,
				  struct policy_handle *connect_handle,
				  struct policy_handle *domain_handle,
				  struct dom_sid2 **domain_sid)
{
	NTSTATUS status;
	WERROR werr;
	struct libnetapi_private_ctx *priv;
	uint32_t resume_handle = 0;
	uint32_t num_entries = 0;
	struct samr_SamArray *sam = NULL;
	const char *domain_name = NULL;
	struct lsa_String lsa_domain_name;
	bool domain_found = true;
	int i;

	priv = talloc_get_type_abort(mem_ctx->private_data,
		struct libnetapi_private_ctx);

	if (is_valid_policy_hnd(&priv->samr.connect_handle)) {
		if ((priv->samr.connect_mask & connect_mask) == connect_mask) {
			*connect_handle = priv->samr.connect_handle;
		} else {
			libnetapi_samr_close_connect_handle(mem_ctx,
				&priv->samr.connect_handle);
		}
	}

	if (is_valid_policy_hnd(&priv->samr.domain_handle)) {
		if ((priv->samr.domain_mask & domain_mask) == domain_mask) {
			*domain_handle = priv->samr.domain_handle;
		} else {
			libnetapi_samr_close_domain_handle(mem_ctx,
				&priv->samr.domain_handle);
		}
	}

	if (priv->samr.domain_sid) {
		*domain_sid = priv->samr.domain_sid;
	}

	if (is_valid_policy_hnd(&priv->samr.connect_handle) &&
	    ((priv->samr.connect_mask & connect_mask) == connect_mask) &&
	    is_valid_policy_hnd(&priv->samr.domain_handle) &&
	    (priv->samr.domain_mask & domain_mask) == domain_mask) {
		return WERR_OK;
	}

	if (!is_valid_policy_hnd(connect_handle)) {
		status = rpccli_try_samr_connects(pipe_cli, mem_ctx,
						  connect_mask,
						  connect_handle);
		if (!NT_STATUS_IS_OK(status)) {
			werr = ntstatus_to_werror(status);
			goto done;
		}
	}

	status = rpccli_samr_EnumDomains(pipe_cli, mem_ctx,
					 connect_handle,
					 &resume_handle,
					 &sam,
					 0xffffffff,
					 &num_entries);
	if (!NT_STATUS_IS_OK(status)) {
		werr = ntstatus_to_werror(status);
		goto done;
	}

	for (i=0; i<num_entries; i++) {

		domain_name = sam->entries[i].name.string;

		if (strequal(domain_name, builtin_domain_name())) {
			continue;
		}

		domain_found = true;
		break;
	}

	if (!domain_found) {
		werr = WERR_NO_SUCH_DOMAIN;
		goto done;
	}

	init_lsa_String(&lsa_domain_name, domain_name);

	status = rpccli_samr_LookupDomain(pipe_cli, mem_ctx,
					  connect_handle,
					  &lsa_domain_name,
					  domain_sid);
	if (!NT_STATUS_IS_OK(status)) {
		werr = ntstatus_to_werror(status);
		goto done;
	}

	status = rpccli_samr_OpenDomain(pipe_cli, mem_ctx,
					connect_handle,
					domain_mask,
					*domain_sid,
					domain_handle);
	if (!NT_STATUS_IS_OK(status)) {
		werr = ntstatus_to_werror(status);
		goto done;
	}

	priv->samr.cli			= pipe_cli;

	priv->samr.domain_name		= domain_name;
	priv->samr.domain_sid		= *domain_sid;

	priv->samr.connect_mask		= connect_mask;
	priv->samr.connect_handle	= *connect_handle;

	priv->samr.domain_mask		= domain_mask;
	priv->samr.domain_handle	= *domain_handle;

	werr = WERR_OK;

 done:
	return werr;
}

/****************************************************************
****************************************************************/

WERROR libnetapi_samr_open_builtin_domain(struct libnetapi_ctx *mem_ctx,
					  struct rpc_pipe_client *pipe_cli,
					  uint32_t connect_mask,
					  uint32_t builtin_mask,
					  struct policy_handle *connect_handle,
					  struct policy_handle *builtin_handle)
{
	NTSTATUS status;
	WERROR werr;
	struct libnetapi_private_ctx *priv;

	priv = talloc_get_type_abort(mem_ctx->private_data,
		struct libnetapi_private_ctx);

	if (is_valid_policy_hnd(&priv->samr.connect_handle)) {
		if ((priv->samr.connect_mask & connect_mask) == connect_mask) {
			*connect_handle = priv->samr.connect_handle;
		} else {
			libnetapi_samr_close_connect_handle(mem_ctx,
				&priv->samr.connect_handle);
		}
	}

	if (is_valid_policy_hnd(&priv->samr.builtin_handle)) {
		if ((priv->samr.builtin_mask & builtin_mask) == builtin_mask) {
			*builtin_handle = priv->samr.builtin_handle;
		} else {
			libnetapi_samr_close_builtin_handle(mem_ctx,
				&priv->samr.builtin_handle);
		}
	}

	if (is_valid_policy_hnd(&priv->samr.connect_handle) &&
	    ((priv->samr.connect_mask & connect_mask) == connect_mask) &&
	    is_valid_policy_hnd(&priv->samr.builtin_handle) &&
	    (priv->samr.builtin_mask & builtin_mask) == builtin_mask) {
		return WERR_OK;
	}

	if (!is_valid_policy_hnd(connect_handle)) {
		status = rpccli_try_samr_connects(pipe_cli, mem_ctx,
						  connect_mask,
						  connect_handle);
		if (!NT_STATUS_IS_OK(status)) {
			werr = ntstatus_to_werror(status);
			goto done;
		}
	}

	status = rpccli_samr_OpenDomain(pipe_cli, mem_ctx,
					connect_handle,
					builtin_mask,
					CONST_DISCARD(DOM_SID *, &global_sid_Builtin),
					builtin_handle);
	if (!NT_STATUS_IS_OK(status)) {
		werr = ntstatus_to_werror(status);
		goto done;
	}

	priv->samr.cli			= pipe_cli;

	priv->samr.connect_mask		= connect_mask;
	priv->samr.connect_handle	= *connect_handle;

	priv->samr.builtin_mask		= builtin_mask;
	priv->samr.builtin_handle	= *builtin_handle;

	werr = WERR_OK;

 done:
	return werr;
}

/****************************************************************
****************************************************************/

void libnetapi_samr_close_domain_handle(struct libnetapi_ctx *ctx,
					struct policy_handle *handle)
{
	struct libnetapi_private_ctx *priv;

	if (!is_valid_policy_hnd(handle)) {
		return;
	}

	priv = talloc_get_type_abort(ctx->private_data,
		struct libnetapi_private_ctx);

	if (!policy_hnd_equal(handle, &priv->samr.domain_handle)) {
		return;
	}

	rpccli_samr_Close(priv->samr.cli, ctx, handle);

	ZERO_STRUCT(priv->samr.domain_handle);
}

/****************************************************************
****************************************************************/

void libnetapi_samr_close_builtin_handle(struct libnetapi_ctx *ctx,
					 struct policy_handle *handle)
{
	struct libnetapi_private_ctx *priv;

	if (!is_valid_policy_hnd(handle)) {
		return;
	}

	priv = talloc_get_type_abort(ctx->private_data,
		struct libnetapi_private_ctx);

	if (!policy_hnd_equal(handle, &priv->samr.builtin_handle)) {
		return;
	}

	rpccli_samr_Close(priv->samr.cli, ctx, handle);

	ZERO_STRUCT(priv->samr.builtin_handle);
}

/****************************************************************
****************************************************************/

void libnetapi_samr_close_connect_handle(struct libnetapi_ctx *ctx,
					 struct policy_handle *handle)
{
	struct libnetapi_private_ctx *priv;

	if (!is_valid_policy_hnd(handle)) {
		return;
	}

	priv = talloc_get_type_abort(ctx->private_data,
		struct libnetapi_private_ctx);

	if (!policy_hnd_equal(handle, &priv->samr.connect_handle)) {
		return;
	}

	rpccli_samr_Close(priv->samr.cli, ctx, handle);

	ZERO_STRUCT(priv->samr.connect_handle);
}

/****************************************************************
****************************************************************/

void libnetapi_samr_free(struct libnetapi_ctx *ctx)
{
	struct libnetapi_private_ctx *priv;

	if (!ctx->private_data) {
		return;
	}

	priv = talloc_get_type_abort(ctx->private_data,
		struct libnetapi_private_ctx);

	libnetapi_samr_close_domain_handle(ctx, &priv->samr.domain_handle);
	libnetapi_samr_close_builtin_handle(ctx, &priv->samr.builtin_handle);
	libnetapi_samr_close_connect_handle(ctx, &priv->samr.connect_handle);
}

File: release/src/router/samba36/source4/scripting/python/examples/samr.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Unix SMB/CIFS implementation.
# Copyright © Jelmer Vernooij <jelmer@samba.org> 2008
#
# Based on samr.js © Andrew Tridgell <tridge@samba.org>
#   
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#   
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#   
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

import sys

sys.path.insert(0, "bin/python")

from samba.dcerpc import samr, security

def display_lsa_string(str):
    return str.string

def FillUserInfo(samr, dom_handle, users, level):
    """fill a user array with user information from samrQueryUserInfo"""
    for i in range(len(users)):
        user_handle = samr.OpenUser(handle, security.SEC_FLAG_MAXIMUM_ALLOWED, users[i].idx)
        info = samr.QueryUserInfo(user_handle, level)
        info.name = users[i].name
        info.idx  = users[i].idx
        users[i] = info
        samr.Close(user_handle)

def toArray((handle, array, num_entries)):
    ret = []
    for x in range(num_entries):
        ret.append((array.entries[x].idx, array.entries[x].name))
    return ret
    

def test_Connect(samr):
    """test the samr_Connect interface"""
    print "Testing samr_Connect"
    return samr.Connect2(None, security.SEC_FLAG_MAXIMUM_ALLOWED)

def test_LookupDomain(samr, handle, domain):
    """test the samr_LookupDomain interface"""
    print "Testing samr_LookupDomain"
    return samr.LookupDomain(handle, domain)

def test_OpenDomain(samr, handle, sid):
    """test the samr_OpenDomain interface"""
    print "Testing samr_OpenDomain"
    return samr.OpenDomain(handle, security.SEC_FLAG_MAXIMUM_ALLOWED, sid)

def test_EnumDomainUsers(samr, dom_handle):
    """test the samr_EnumDomainUsers interface"""
    print "Testing samr_EnumDomainUsers"
    users = toArray(samr.EnumDomainUsers(dom_handle, 0, 0, -1))
    print "Found %d users" % len(users)
    for idx, user in users:
        print "\t%s\t(%d)" % (user.string, idx)

def test_EnumDomainGroups(samr, dom_handle):
    """test the samr_EnumDomainGroups interface"""
    print "Testing samr_EnumDomainGroups"
    groups = toArray(samr.EnumDomainGroups(dom_handle, 0, 0))
    print "Found %d groups" % len(groups)
    for idx, group in groups:
        print "\t%s\t(%d)" % (group.string, idx)

def test_domain_ops(samr, dom_handle):
    """test domain specific ops"""
    test_EnumDomainUsers(samr, dom_handle)
    test_EnumDomainGroups(samr, dom_handle)

def test_EnumDomains(samr, handle):
    """test the samr_EnumDomains interface"""
    print "Testing samr_EnumDomains"

    domains = toArray(samr.EnumDomains(handle, 0, -1))
    print "Found %d domains" % len(domains)
    for idx, domain in domains:
        print "\t%s (%d)" % (display_lsa_string(domain), idx)
    for idx, domain in domains:
        print "Testing domain %s" % display_lsa_string(domain)
        sid = samr.LookupDomain(handle, domain)
        dom_handle = test_OpenDomain(samr, handle, sid)
        test_domain_ops(samr, dom_handle)
        samr.Close(dom_handle)

if len(sys.argv) != 2:
   print "Usage: samr.js <BINDING>"
   sys.exit(1)

binding = sys.argv[1]

print "Connecting to %s" % binding
try:
    samr = samr.samr(binding)
except Exception, e:
    print "Failed to connect to %s: %s" % (binding, e.message)
    sys.exit(1)

handle = test_Connect(samr)
test_EnumDomains(samr, handle)
samr.Close(handle)

print "All OK"
